core/desugarInners.ml

```diff
@@ -172,7 +172,12 @@ object (o : 'self_type)
                let o = o#with_visiting (StringSet.add (Binder.to_name rec_binder) visiting_funs) in
                let (o, tyvars) = o#quantifiers tyvars in
                let (o, inner) = o#datatype inner in
-               let inner_effects = TransformSugar.fun_effects inner (fst lam) in
+               let lam_in =
```

Comment:
Take this as a comment rather than an explicit request for changes, but I notice this pattern a lot: knowing that something must be a `NormalFunLit`, and then matching, and asserting false in the impossible case. You may want to abstract over this by moving this logic to `Sugartypes` by having a function there:

```
let get_normal_funlit = function
  | NormalFunLit x -> x
  | _ -> internal_error "SwitchFunLit where NormalFunLit expected"
```

and then you could just call it as

`let lam_in = get_normal_funlit lam in ...`

and save yourself the unwrapping effort at the callsites.